### DISCUSSION TITLE
Timezone issue

### DISCUSSION THREAD

**Discussion by ssljivic** [mikro-orm/mikro-orm#5591] [General]
### Describe the bug

Mikro ORM applies negative timezone offset when reading the Postgres timestamp (without tz) values which creates invalid  date objects.

If option `forceUtcTimezone` is set to `true`, the conversion is not applied, but since PG column is timestamp (without tz), this should work correctly without `forceUtcTimezone` applied.

I am in GMT+2 TZ and the below test was ran at "2024-05-22T14:50:28.488" local time. The test output is:

```
DATE     : 2024-05-22T12:50:28.488Z
ORM DATE : 2024-05-22T10:50:28.488Z
RAW DATE : 2024-05-22 12:50:28.488
```

### Reproduction

```
import { Entity, type Opt, PrimaryKey, Property, ReflectMetadataProvider } from '@mikro-orm/core'
import { LoadStrategy, MikroORM, PostgreSqlDriver } from '@mikro-orm/postgresql'

@Entity()
export class Test {
  @PrimaryKey()
  id!: number

  @Property()
  name!: string

  @Property({ type: 'timestamp', length: 0, defaultRaw: 'now()' })
  createdAt: Date & Opt = new Date()
}

describe('MikroORM', () => {
  let orm: MikroORM

  beforeAll(async () => {
    orm = await MikroORM.init({
      host: 'localhost',
      port: 5432,
      user: 'mikro-orm-test',
      password: 'mikro-orm-test',
      dbName: 'mikro-orm-test',

      driver: PostgreSqlDriver,

      loadStrategy: LoadStrategy.JOINED,
      // forceUtcTimezone: true,
      metadataProvider: ReflectMetadataProvider,
      allowGlobalContext: false,

      debug: true,
      entities: [Test],
    })
  })

  afterAll(async () => {
    await orm.close(true)
  })

  it('should handle dates correctly', async () => {
    let em = orm.em.fork()

    await em.getKnex().raw(`
      create table if not exists test
      (
        id         serial primary key,
        name       varchar(255)               not null,
        created_at timestamp(3) default now() not null
      );

      delete from test;
    `)

    const name = 'john'
    const createdAt = new Date()

    const t1 = new Test()
    t1.name = name
    t1.createdAt = createdAt

    await em.insert(t1)
    await em.flush()

    em = em.fork()
    const res = await em.findOneOrFail(Test, { name })

    const rawRes = await em.getKnex().raw('select * from test where name = :name', { name })

    console.log('DATE     :', createdAt.toISOString())
    console.log('ORM DATE :', res.createdAt.toISOString())
    console.log('RAW DATE :', rawRes.rows[0].created_at)

    expect(res.createdAt.getTime()).toBe(createdAt.getTime())
  })
})
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.2.5

### Node.js version

18.17.1

### Operating system

MacOS 14.0

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ssljivic** [comment]
I am not sure why is the original issue converted to discussion.
I would say that this is the error, because writing one date value and then reading it from the Db, produces a completely different timestamp.

**Reply by gregoryorton-ws** [reply]
I think this is a repro - https://github.com/mikro-orm/mikro-orm/discussions/5891

**Comment by ssljivic** [comment]
There is definitely some very strange issue here. If `forceUtcTimezone` is set to `false`, when I insert date `2024-05-22T14:50:28.488` (GMT+2), and I fetch that value from DB, I would expect to get the `2024-05-22T14:50:28.488` (GMT+2). Instead, I am getting date that is shifted for 4 hours.

**Reply by ssljivic** [reply]
If I do the same insert/select using Kysely or Knex, everything works as expected.

**Reply by B4nan** [reply]
The way things are set up, you shouldn't get a `Date` instance out of knex at all, you should only get the raw value as string (your repro confirms that too). So comparing the value to the `Date` objects is a bit weird on its own.

**Comment by B4nan** [comment]
I will have a closer look at the repro, it looks a bit weird, but those issues are often about the environment and not the code itself. Overall, the most sane approach is to enforce UTC everywhere.

> If option forceUtcTimezone is set to true, the conversion is not applied, but since PG column is timestamp (without tz), this should work correctly without forceUtcTimezone applied.

All the timezone options are there _only_ for the column types that don't hold the TZ - it's never about the timezone-aware types.

**Comment by gregoryorton-ws** [comment]
@ssljivic @B4nan   - I am also seeing this bug, and it is a bug! 

I have set my postgreSQL columns in my model to be `timestamp` not the default `timestamptz`:

```ts
  // ...

 @Property({
    fieldName: 'starts_at',
    columnType: 'timestamp',
    nullable: false,
  })
  declare startsAt: Date;

  @Property({
    fieldName: 'ends_at',
    columnType: 'timestamp',
    nullable: false,
  })
  declare endsAt: Date;
  
  // ...
```

I use dayJS to make sure I'm producing UTC times before coverting to JS date (`dayjs.utc(date string).toDate()`) and then store in database.) When I store the value I can see it's the UTC value in the database. (no change in the Date value vs. value persisted)

![CleanShot 2024-08-02 at 13 53 00](https://github.com/user-attachments/assets/83436a29-5847-4812-8dcb-4a7835ce5878)

But when I use mikroORM in integration tests to retrieve the value I can see v. clearly that MikroORM is "converting" the value to what it thinks is UTC time based on the timezone of my system/database....

Example:

```ts
let managedRecord = this.repo.findOneOrFail(managedRecord)
const endsAt = dayjs.utc().startOf('day').hour('18').toDate() // produces 2024-08-01 18:00:00
managedRecord.assign({ endsAt }) 
this.repo.getEntityManager().flush()
managedRecord = this.repo.findOneOrFail(managedRecord)

expect(endsAt.toISOString()).toEqual(managedRecord.toObject().endsAt.toISOString()) 
// fails with `Expected: 2024-08-01T18:00:00.000Z Received: 2024-08-01T10:00:00.000Z`  !!!!!!!!! WRONG WRONG WRONG !!!!!!!

```



--- 

extra info:

I work in Asia which means my timezone is `Asia/Shanghai` and my postgreSQL server conf is also in this timezone, which currently is UTC+8 or GMT+7 (it's BST so 7 hours right now)

```
timezone = 'Asia/Shanghai'
```

**Comment by jefmenegazzo** [comment]
@B4nan @gregoryorton-ws @ssljivic   I have also been experiencing this bug using the timestamp field (timestamp without timezone) in Postgres.

There is a bug in the use of the configuration flags forceUtcTimezone and timezone. From what I looked at in the repository code, I understood the following:

- When `forceUtcTimezone = false` is used, it will consider the `timezone`.
- When `forceUtcTimezone = true` is used, it will consider the `timezone = "Z"`.

There are two scenarios:

**1. Save the timestamp in UTC in the database:**

In this case, using `forceUtcTimezone = true` and `timezone = "-03:00" `(I am in GMT-3), everything works correctly and the offsets are applied:

| Value in JS  before save in DB | Value in DB after saved | Value in JS after read from DB|
| -------- | ------- | -------|
| 2024-18-02 09:00:00 GMT−3 | 2024-18-02 12:00:00 | 2024-18-02 09:00:00 GMT−3 |

**2. Save the timestamp in the local timezone in the database:**

In this case, using `forceUtcTimezone = false` because I want to save the datetime in the local timezone, the `timezone` flag will be ignored:

| Value in JS  before save in DB | Value in DB after saved | Value in JS after read from DB|
| -------- | ------- | -------|
| 2024-18-02 09:00:00 GMT−3 | 2024-18-02 12:00:00 | 2024-18-02 12:00:00 GMT−3 |

Here is the bug: **even with `forceUtcTimezone = false`, MikroOrm is always saving in UTC.**

**Reply by jefmenegazzo** [reply]
@B4nan @gregoryorton-ws I don't think this is a good reproduction of the bug that @ssljivic and I are experiencing. I will provide a repro later, but in my case and @ssljivic's, it is indeed a bug. In your case, @gregoryorton-ws, it seems to be just a configuration error. Have you tried setting `forceUtcTimezone = true` and your timezone to `timezone = "+08:00"` in `MikroORM.init`? In this case, you don't need to manually convert to UTC, leaving that to MikroORM (considering that in your system, the timestamp dates can be stored in UTC in database, as described in scenario 1).

**Reply by gregoryorton-ws** [reply]
umm, not sure why you say that @jefmenegazzo - I've never met an ORM I need to say which timezone I'm working in - TypeORM, Prisma.

Also, `forceUtcTimezone` as far as I can tell, simply forces all column types in DB to be `timestamp` instead of `timestamptz`. Nothing else. 

> In this case, you don't need to manually convert to UTC, leaving that to MikroORM (considering that in your system, the timestamp dates can be stored in UTC in database, as described in scenario 1).

I'm not sure what you're talking about, really. I don't manually convert to UTC anyway? JS dates have no concept of timezone. They're only transformed to local date time when you say `.toLocaleString()`. 

A `timestamp` column is stored as a unix epoch timestamp. a `timestamptz` column is also stored as unix epoch timestamp.

The only difference between `timestamp` and `timestamptz` is that if you try to insert a string like `1970-01-01T18:00:00.000+08:00` it will first convert it to UTC string `1970-01-01T10:00:00.000Z` and store it. 

The only reason I use `dayjs.utc("2024-08-01").startOf("day").hour(17).toDate()` is to ensure that nothing about my local date will impact the test. This produces very clear JS date which resolves to an ISO string of `2024-08-01T17:00:00.000Z`

Regardless of my timezone, this is the string that should be passed to the JS date constructor `new Date("2024-08-01T17:00:00.000Z")` when retrieved from the database (for a `timestamp` type column). It's that simple. 

**I suspect the bug is something like this:**

Because it's a `timestamp` column, it's assumed that the JS date is should be constructed without the `Z` part of the string, i.e. `new Date("2024-08-01T17:00:00.000")`

this will create the JS date object in the same timezone as your system... simply proved by typing this in the console:

![CleanShot 2024-08-06 at 00 12 53](https://github.com/user-attachments/assets/af2417c6-85bd-4fc1-9ae2-305aa5c5ef9f)

Printing this date as UTC ISO String with `date.toISOString()` produces the same error I see...

![CleanShot 2024-08-06 at 00 15 30](https://github.com/user-attachments/assets/a858709e-0659-45be-9693-039a52dece51)

**Reply by B4nan** [reply]
It's not implemented this way in the ORM, `timestamp` column returns a string value without the timezone information, not a number (such a number would be considered a UTC timestamp), and by default, the ORM will use the local timezone to parse it - it will just do `new Date(valueFromDatabase)`. I doubt other ORMs would be doing something else. With `forceUtcTimezone` you make the ORM add the `Z` suffix there, parsing the dates as UTC. Using that option is equivalent to `timezone: 'Z'`.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/EntityComparator.ts#L324-L339

**Reply by gregoryorton-ws** [reply]
> It's not implemented this way in the ORM, timestamp column returns a string value without the timezone information, not a number (such a number would be considered a UTC timestamp)

Yes @B4nan - when I said "it's a unix epoch timestamp" I meant the internal representation. I can tell you with 100% certainty, neither `timestamp` or `timestampz` actually store ANY information about timezone in the internal representation of the value. 

> and by default, the ORM will use the local timezone to parse it - it will just do new Date(valueFromDatabase)

and that's what I'm saying is an issue. You get back from the db the string `YYYY-MM-DD HH:mm:ss` from the database without the `z` part and just create a date... this doesn't make sense to me.

A `timestamp` column is UTC value and should be produced in JS date object as UTC value by adding the `Z` if it comes from a column called `timestamp` - you know a `timestamp` type is timezone-less and therefore is a  UTC value!  Why then create it in local time?! the `forceUtcTimezone` setting is an unnecessary burden/complication. It only makes sense for `timestamptz` columns.

>  I doubt other ORMs would be doing something else.

I just finished two other projects using prisma and had none of these issues and no special configuration needed! Same Postgres server with same local timezone of `Asia/Shanghai` and using `timestamp` columns. It also ONLY returns JS dates for date columns (you can't even return strings!) and there were just no issues. The date created was created from db value was always correct.

**Reply by B4nan** [reply]
Hmm, I agree that this feels like a bug, especially because it works fine in the other dialects without the `forceUtcTimezone ` option. It's all caused by the underlying postgres driver returning a string date instead of a number as all the other dialects are doing for timestamp columns. I can imagine prisma and its rust-based engine will also use numbers, they don't depend on the `pg` library afaik. This all feels like a weird quirk from the `pg` library to me. I just tried to enable the default type mapping in there (which means mapping it to `Date` on their part) and it also returns the wrong date with the timezone shift.

**Reply by gregoryorton-ws** [reply]
Well, in mySQL they have the `datetime` column which is definitely TZ-free but is not a number (stored internally it probably is). You have to assume that's also UTC right? Things could get very messy otherwise... 😂

I think we'll have to forge ahead with `forceUtcTimezone` for now. Thanks for taking time to investigate and discuss. Personally I think that the `Z` should be added for all `timestamp` type  columns when constructing the JS date, and the `timezone` setting for the ORM should only take effect for `timetstamptz` columns as per the design of retrieving based on connection...

**Reply by B4nan** [reply]
> and the timezone setting for the ORM should only take effect for timetstamptz columns as per the design of retrieving based on connection...

This makes no sense, `timestamptz` columns are timezone aware, the value the ORM gets from the database includes the correct timezone, none of this is relevant for those values, the ORM never overrides a timezone, these options are about working with values _without the timezone_. You keep talking about how the values are stored, but this is not about storage, its about interpretation. One value is returned as a datetime string without tz, one is returned with it.

**Reply by gregoryorton-ws** [reply]
> the ORM never overrides a timezone

But that's exactly what you're doing with current behavior! timestamps are by definition UTC values! whether or not the database returns a number or returns a string, the `timestamp` column is UTC only. It's down to the user to do any conversion post retrieval, not an ORM setting!

> these options are about working with values without the timezone.

then they are pointless options. 

> the value the ORM gets from the database includes the correct timezone.

No, not `correct timezone` -  **configured timezone** - the timezone it's told to use! Either:

1) the timezone in the database config at startup...
2) the timezone specified in the session parameter (e.g. `set timezone TO 'CST';`)

**I was making an assumption that your `timezone` setting in the ORM config would be for 2) - the session settings for the db connection it would make.  Because specifying a timezone for `timestamp` column is just seems weird and confusing...**

> the ORM never overrides a timezone

full circle - no, but it can specify the timezone it wants the db to use for `timestamptz` column retrieval via sessions/drive connection settings...

**Reply by gregoryorton-ws** [reply]
> timestamptz columns are timezone aware

I don't meant to confuse you... just ask the question: "Aware of what timezone?" - 

They're aware of ...the timezone of the database config or the current session context...that's a setting I would maybe expect ORM to have... not settings to  mess with how `timestamp` columns are parsed...

> You keep talking about how the values are stored, but this is not about storage,

Only because I'm trying to force the message that the timezone is a relative thing based on connection...

**Reply by gregoryorton-ws** [reply]
One more example. I add a new column to my `test` table. this time a `timestamptz` column.

```sql
CREATE TABLE "public"."test" (
    "id" int4 NOT NULL DEFAULT nextval('test_id_seq'::regclass),
    "utc_only_please" timestamp,
    "tz_only_please" timestamptz,
    PRIMARY KEY ("id")
);
```

one test row:

```
INSERT INTO "public"."test" ("id", "utc_only_please", "tz_only_please") VALUES
(9, '2024-08-01 12:00:00', '2024-08-01 12:00:00Z');
```

---

set timezone to asia and read...

```
SET timezone to 'Asia/Shanghai';
SELECT * FROM `test`;

 id |   utc_only_please   |     tz_only_please
----+---------------------+------------------------
  9 | 2024-08-01 12:00:00 | 2024-08-01 20:00:00+08
```

---

set timezone to EU...

```
SET timezone to 'Europe/Berlin'
SELECT * FROM `test`;
 id |   utc_only_please   |     tz_only_please
----+---------------------+------------------------
  9 | 2024-08-01 12:00:00 | 2024-08-01 14:00:00+02
```

---

IMHO 'timezone' settings passed to ORM should not affect the `timestamp` columns... only the selected timezone used for retrieving `timestampz` info.

**Reply by B4nan** [reply]
You keep talking about how the database works, but the `timezone` option in the ORM is not the same as you use in those examples (the ORM never calls the `set timezone to ...` query), its purpose is the very opposite, to provide the missing timezone for `timestamp` column, and this option is bound to the `forceUtcTimezone`. As such, it cannot have any effect on the `timestamptz` column, their values are coming from the database including the timezone (which your snippets only confirm), and the ORM respects that.

So please stop trying to convince me how the ORM should work internally when you don't understand what the options are supposed to be doing. We can't just change the meaning of those options, not outside of major versions. I've already agreed that the behavior in postgres is wrong, and I am open to discussion how to narrow all of this, but the way you communicate here is really tiring and I am not interested in such conversations really.

**Reply by gregoryorton-ws** [reply]
> You keep talking about how the database works, 

Because I'm trying to communicate that I think that the ORM should follow how the database works and right now I think it interferes with the way postgreSQL works... - so I'll give it one more try and hope it's not "tiring". If you still think I'm talking BS, then I'll just give up and try not to bother you with ideas or suggestions again.

>  the timezone option in the ORM is not the same as you use in those examples (the ORM never calls the set timezone to ... query), ... and this option is bound to the forceUtcTimezone

Yes, I get it. I have parsed, understood, and internalized it. You've explained it several times. What you're failing to internalize  (I appreciate I might be not communicating so well myself) _is I don't care how that's how they work now._ I find the two settings unnecessary and confusing and what I'm trying to do is convince you to change it and radically depart from it.  

> its purpose is the very opposite, to provide the missing timezone for timestamp column, 

And this is the crux of it. I don't know the history of the project or how the design of these "options" were arrived at, but _it's this idea of "provide the missing timezone for timestamp column" is EXACTLY what  I think is incorrect and is what I'm trying to get you to depart from!_ 

You went to great pains to suggest that an ORM shouldn't interfere or change the values which come from the database, yet you have decreed there is "a missing timezone" and it's user configurable? 🤔. 


The value returned (because of the driver, whatever) could be  a string, or a number, or a ...- what matters is value is derived from a stored timestamp... Timestamps by definition are not TZ-aware, but most importantly are considered in the UTC timezone. **There IS NO MISSING TIMEZONE**

IMO for `timestamp` columns, if a user specifies run time type as:
- string: return the raw string from the database (or even better, as you know it's a timestamp, return the UTC ISO string with a "Z")
- number: return the unix timestamp
- Date: return the Date constructed with the UTC ISO so that it remains the SAME value, not some TZ setting from the ORM...


**If you still think I'm misunderstanding what these options are doing, then IMO this is even more evidence they shouldn't exist or they shouldn't be exposed (which is my suggestion) - internally everything should be done to return a UTC date of a timezone-unaware column!**


>  As such, it cannot have any effect on the timestamptz column, their values are coming from the database including the timezone 

And I have understood that... 

> So please stop trying to convince me how the ORM should work internally when you don't understand what the options are supposed to be doing.

Except I do. What I've obviously been failing to communicate and convince is I want you to throw away the settings of `timezone` and or `forceUtcTimezone` on `timestamp` columns because I see them as unnecessary and confusing.

> We can't just change the meaning of those options, not outside of major versions.

And I'm fine with this. If you/users/patrons/contributors want to keep some notion of control over timezone settings for specific column types in specific dialects then "huzzah"? 

I am asking that the radical change comes under consideration in the next major version for the purposes of simplicity and easy-to-reason-about features. 

IMO I think it's a better approach and I've given you a sane suggestion on how to use a `timezone` setting in a more intuitive way. (i.e. use it to control the session timezone for the connection for timezone-aware columns!) 

>  I've already agreed that the behavior in postgres is wrong,

Sure. I know that there are other dialects, (my experience is broadly around MariaDb/MySQL/PostgreSQL, less MSSQL), so I know each DB supports different types, but as I suggested elsewhere, I would use the same "only UTC" behaviour for `Datetime` fields in MySQL. They're not TZ aware so the only sane way to deal with them is treat them as UTC and only ever UTC....

> and I am open to discussion how to narrow all of this,

Which is what I've been trying to communicate. You might think I'm being tiring or combative - I don't know how to adjust it at this point. I already said I'd accept using `forceUtcTimezone` as a workaround, but I thought we'd moved past that. I want to provide constructive feedback about how to improve the DX/usefullness in a way that makes more intuitive sense...

Anyway, best of luck with Mikro-ORM development for the future! It's really a great library and a fantastic achievement. It's still miles ahead of TypeORM, Sequelize, and Prisma (yes, I've used them all on various projects) so thank you for what you've already done.

**Comment by B4nan** [comment]
Looks like the culprit to all of this is the fact that postgres driver always uses UTC when writing data, but interprets them in local timezone when reading (unless you use `forceUtcTimezone`):

https://github.com/mikro-orm/mikro-orm/blob/master/packages/postgresql/src/PostgreSqlPlatform.ts#L278
https://github.com/mikro-orm/mikro-orm/blob/master/packages/postgresql/src/PostgreSqlPlatform.ts#L387

When I adjust that to respect the `timezone` option, both reproductions from here are working, as in fetching the same date as it was stored. The stored value is in local timezone unless you use `forceUtcTimezone`, using that turns them into UTC. So to fix this, its actually enough to [remove this line](https://github.com/mikro-orm/mikro-orm/blob/master/packages/postgresql/src/PostgreSqlPlatform.ts#L278), and let the value be converted by the parent implementation which respects the timezone. 

I also checked the other dialects, and it's quite a mess:

- sqlite converts to number right ahead, so uses UTC and never respects the `timezone` option when parsing, since the values from db are also numeric timestamps
- MySQL and MSSQL stores values in local timezone, and respects the timezone when reading too, with `forceUtcTimezone` the value is stored in UTC
- postgres always writes UTC and reads based on the `timezone` option as mentioned earlier

We should definitely clean this up, since some drivers operatate by default as if `forceUtcTimezone` is true, while some use a local timezone by default (and then there is postgres in the middle). But it will be breaking, we can't just change how things are interpreted in the other dialects, so that needs to wait for v7. 

Now the question is what should be the default and what to do with that `timezone` option we have now, since I agree its really weird. I am thinking about setting `forceUtcTimezone` to true by default in v7, and either removing the `timezone` option completely, or turning that into the SQL `set timezone to ...` executed on every new connection.

For now we should fix the postgres behavior for sure, now it makes no sense, and since we were always storing UTC, it feels "less breaking" to make the postgres driver behave like `forceUtcTimezone` is true, so the same as SQLite drivers do now. MySQL will remain `forceUtcTimezone: false` behavior and we'll swap that to true in v7.

I am thinking about actually having the `forceUtcTimezone` default dependent on the driver for now, since that could be relatively easy refactor and in v7 we would just unify the defaults.

**Reply by B4nan** [reply]
#7087 enables `forceUtcTimezone ` by default (v7)
